File: server/WebApp/Models/Shared/Result.cs
namespace WebApp.Models.Shared;

public class Result
{
    public bool IsSuccess { get; }

    public bool IsFailure => !IsSuccess;

    public Error? Error { get; }

    protected Result(Error error)
    {
        Error = error;
        IsSuccess = false;
    }

    protected Result()
    {
        IsSuccess = true;
    }

    public static Result Success() => new();
    public static Result<T> Success<T>(T value) => new(value);
    public static Result Failure(Error error) => new(error);
    public static Result<T> Failure<T>(Error error) => new(error);
    public static Result<T> Create<T>(T value) => new(value);
}

public class Result<T> : Result
{
    public T? Value { get; }

    protected internal Result(Error error) : base(error)
    {}

    protected internal Result(T value)
    {
        Value = value;
    }
}

File: client/src/primitives/Result.ts
import type { AppError } from './Error'

export class Result<T> {
  public readonly value?: T
  public readonly isSuccess: boolean
  public readonly error?: AppError

  private constructor(value?: T, error?: AppError) {
    if (error) {
      this.error = error
      this.isSuccess = false
    } else {
      this.value = value
      this.isSuccess = true
    }
  }

  public static success<T>(value: T): Result<T> {
    return new Result<T>(value)
  }

  public static failure<T>(error: AppError): Result<T> {
    return new Result<T>(undefined, error)
  }

  public static create<T>(value: T): Result<T> {
    return new Result<T>(value)
  }

  public get isFailure(): boolean {
    return !this.isSuccess
  }
}
